## Association Rule Mining in Retail Store

### Problem Statement:
 * What are the items that may be frequently purchased together?

### Objective:
* To know which items are frequently purchased together, keeping both item together will help to increase sales.


### Introduction
* Association rule mining is one of an important technique of data mining for knowledge discovery.
* The knowledge of the correlation between the items in the data transaction can use association rule mining.
* Retail store analysis is one of an application area of association rule mining technique.
* The possible percentage of the correlation of combined items gives the new knowledge. Therefore, it is a very helpful for determiner to take the decisions

### Analysis

In [225]:
## Importing Required Library

import pandas as pd
import numpy as np
from mlxtend.frequent_patterns import apriori, association_rules

In [226]:
# Reading Excel file 
bread = pd.read_excel('raw_bread.xlsx')

In [227]:
## Here we have transaction data, which include column, Date,Time,Transaction,Item
## we should remove duplicate transaction, it shows quantity of item in same transaction,
# it is not needed in appriori aglo as we only care about different item in particular transaction
bread

Date,Time,Transaction,Item
0              2016-10-30,09:58:11,1,Bread
1       2016-10-30,10:05:34,2,Scandinavian
2       2016-10-30,10:05:34,2,Scandinavian
3      2016-10-30,10:07:57,3,Hot chocolate
4                2016-10-30,10:07:57,3,Jam
...                                    ...
21288      2017-04-09,14:32:58,9682,Coffee
21289         2017-04-09,14:32:58,9682,Tea
21290      2017-04-09,14:57:06,9683,Coffee
21291      2017-04-09,14:57:06,9683,Pastry
21292   2017-04-09,15:04:24,9684,Smoothies

[21293 rows x 1 columns]

In [228]:
## dropping Duplicate Transaction
bread = bread.drop_duplicates()

In [229]:
## we need to split transaction data into Dataframe/tabular structure as follow
new = bread['Date,Time,Transaction,Item'].str.split(',', n = 3, expand = True)

In [230]:
## assigning column to data frame "bread"
bread['Date'] = new[0]
bread['Time'] = new[1]
bread['Transaction'] = new[2]
bread['Item'] = new[3]

C:\Users\91892\anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
C:\Users\91892\anaconda3\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until
C:\Users\91892\anaconda3\lib\site-packages\ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = v

In [242]:
# in this dataframe we only need column Trasaction and Item, rest is not needed in association mining rule
bread

Date,Time,Transaction,Item        Date      Time Transaction  \
0              2016-10-30,09:58:11,1,Bread  2016-10-30  09:58:11           1   
1       2016-10-30,10:05:34,2,Scandinavian  2016-10-30  10:05:34           2   
3      2016-10-30,10:07:57,3,Hot chocolate  2016-10-30  10:07:57           3   
4                2016-10-30,10:07:57,3,Jam  2016-10-30  10:07:57           3   
5            2016-10-30,10:07:57,3,Cookies  2016-10-30  10:07:57           3   
...                                    ...         ...       ...         ...   
21288      2017-04-09,14:32:58,9682,Coffee  2017-04-09  14:32:58        9682   
21289         2017-04-09,14:32:58,9682,Tea  2017-04-09  14:32:58        9682   
21290      2017-04-09,14:57:06,9683,Coffee  2017-04-09  14:57:06        9683   
21291      2017-04-09,14:57:06,9683,Pastry  2017-04-09  14:57:06        9683   
21292   2017-04-09,15:04:24,9684,Smoothies  2017-04-09  15:04:24        9684   

                Item  
0              Bread  
1       Scandinavian  
3      Hot chocolate  
4                Jam  
5            Cookies  
...              ...  
21288         Coffee  
21289            Tea  
21290         Coffee  
21291         Pastry  
21292      Smoothies  

[19640 rows x 5 columns]

In [232]:
# we need to convert cloumn transacton & item into Crosstab or we can say Binary Matrix as follow
tab = pd.crosstab(index= bread['Transaction'], columns= bread['Item'])
tab

Item         Adjustment  Afternoon with the baker  Alfajores  Argentina Night  \
Transaction                                                                     
1                     0                         0          0                0   
10                    0                         0          0                0   
100                   0                         0          0                0   
1000                  0                         0          0                0   
1001                  0                         0          0                0   
...                 ...                       ...        ...              ...   
995                   0                         0          0                0   
996                   0                         0          0                0   
997                   0                         0          0                0   
998                   0                         0          0                0   
999                   0                         0          1                0   

Item         Art Tray  Bacon  Baguette  Bakewell  Bare Popcorn  Basket  ...  \
Transaction                                                             ...   
1                   0      0         0         0             0       0  ...   
10                  0      0         0         0             0       0  ...   
100                 0      0         0         0             0       0  ...   
1000                0      0         0         0             0       0  ...   
1001                0      0         0         0             0       0  ...   
...               ...    ...       ...       ...           ...     ...  ...   
995                 0      0         0         0             0       0  ...   
996                 0      0         0         0             0       0  ...   
997                 0      0         0         0             0       0  ...   
998                 0      0         0         0             0       0  ...   
999                 0      0         0         0             0       0  ...   

Item         The BART  The Nomad  Tiffin  Toast  Truffles  Tshirt  \
Transaction                                                         
1                   0          0       0      0         0       0   
10                  0          0       0      0         0       0   
100                 0          0       0      0         0       0   
1000                0          0       0      0         0       0   
1001                0          0       0      0         1       0   
...               ...        ...     ...    ...       ...     ...   
995                 0          0       0      0         0       0   
996                 0          0       0      0         1       0   
997                 0          0       0      0         0       0   
998                 0          0       0      0         0       0   
999                 0          0       0      0         0       0   

Item         Valentine's card  Vegan Feast  Vegan mincepie  Victorian Sponge  
Transaction                                                                   
1                           0            0               0                 0  
10                          0            0               0                 0  
100                         0            0               0                 0  
1000                        0            0               0                 0  
1001                        0            0               0                 0  
...                       ...          ...             ...               ...  
995                         0            0               0                 0  
996                         0            0               0                 0  
997                         0            0               0                 0  
998                         0            0               0                 0  
999                         0            0               0              

In [249]:
## Just writing cdv file to check result
## I came to know that, we have one unwanted column named "NONE", we should remove it as follow and proceed further
#tab.to_csv('tab.csv')

In [246]:
## removing unwanted col "NONE"
tab = tab.drop(['NONE'], axis = 1)

In [250]:
tab

Item         Adjustment  Afternoon with the baker  Alfajores  Argentina Night  \
Transaction                                                                     
1                     0                         0          0                0   
10                    0                         0          0                0   
100                   0                         0          0                0   
1000                  0                         0          0                0   
1001                  0                         0          0                0   
...                 ...                       ...        ...              ...   
995                   0                         0          0                0   
996                   0                         0          0                0   
997                   0                         0          0                0   
998                   0                         0          0                0   
999                   0                         0          1                0   

Item         Art Tray  Bacon  Baguette  Bakewell  Bare Popcorn  Basket  ...  \
Transaction                                                             ...   
1                   0      0         0         0             0       0  ...   
10                  0      0         0         0             0       0  ...   
100                 0      0         0         0             0       0  ...   
1000                0      0         0         0             0       0  ...   
1001                0      0         0         0             0       0  ...   
...               ...    ...       ...       ...           ...     ...  ...   
995                 0      0         0         0             0       0  ...   
996                 0      0         0         0             0       0  ...   
997                 0      0         0         0             0       0  ...   
998                 0      0         0         0             0       0  ...   
999                 0      0         0         0             0       0  ...   

Item         The BART  The Nomad  Tiffin  Toast  Truffles  Tshirt  \
Transaction                                                         
1                   0          0       0      0         0       0   
10                  0          0       0      0         0       0   
100                 0          0       0      0         0       0   
1000                0          0       0      0         0       0   
1001                0          0       0      0         1       0   
...               ...        ...     ...    ...       ...     ...   
995                 0          0       0      0         0       0   
996                 0          0       0      0         1       0   
997                 0          0       0      0         0       0   
998                 0          0       0      0         0       0   
999                 0          0       0      0         0       0   

Item         Valentine's card  Vegan Feast  Vegan mincepie  Victorian Sponge  
Transaction                                                                   
1                           0            0               0                 0  
10                          0            0               0                 0  
100                         0            0               0                 0  
1000                        0            0               0                 0  
1001                        0            0               0                 0  
...                       ...          ...             ...               ...  
995                         0            0               0                 0  
996                         0            0               0                 0  
997                         0            0               0                 0  
998                         0            0               0                 0  
999                         0            0               0              

In [304]:
## finding associate(item) which brought together with frequency greater than 0.1%
associate = apriori(df = tab, min_support= 0.001, use_colnames= True)

In [305]:
## Sorting Associate by support in ascending order.
associate.sort_values(by = 'support').head()

support                       itemsets
470  0.001049  (Tea, Cake, Sandwich, Coffee)
313  0.001049    (Alfajores, Juice, Cookies)
420  0.001049   (Pastry, Farm House, Coffee)
315  0.001049       (Tea, Pastry, Alfajores)
71   0.001049    (Spanish Brunch, Alfajores)

In [306]:
## Createing associate rule such that item brought with conditional probability(Confidence) more than 50% with corresponding item
asso_rule = association_rules(associate, min_threshold= 0.7)

In [307]:
asso_rule.head()

antecedents consequents  antecedent support  consequent support  \
0  (Extra Salami or Feta)    (Coffee)            0.003987            0.475081   
1      (Keeping It Local)    (Coffee)            0.006610            0.475081   
2                 (Toast)    (Coffee)            0.033365            0.475081   
3           (Cake, Salad)    (Coffee)            0.001364            0.475081   
4           (Cake, Toast)    (Coffee)            0.002203            0.475081   

    support  confidence      lift  leverage  conviction  
0  0.003253    0.815789  1.717158  0.001358    2.849559  
1  0.005351    0.809524  1.703969  0.002211    2.755823  
2  0.023502    0.704403  1.482699  0.007651    1.775789  
3  0.001049    0.769231  1.619156  0.000401    2.274648  
4  0.001574    0.714286  1.503502  0.000527    1.837215

In [308]:
asso_rule.sort_values(by='lift', ascending= False).head(10)

antecedents consequents  antecedent support  \
7   (Salad, Extra Salami or Feta)    (Coffee)            0.001679   
11                (Pastry, Toast)    (Coffee)            0.001574   
8   (Hearty & Seasonal, Sandwich)    (Coffee)            0.001469   
5          (Vegan mincepie, Cake)    (Coffee)            0.001259   
12              (Sandwich, Salad)    (Coffee)            0.001889   
0          (Extra Salami or Feta)    (Coffee)            0.003987   
1              (Keeping It Local)    (Coffee)            0.006610   
6                (Cookies, Scone)    (Coffee)            0.001993   
9                 (Pastry, Juice)    (Coffee)            0.002308   
3                   (Cake, Salad)    (Coffee)            0.001364   

    consequent support   support  confidence      lift  leverage  conviction  
7             0.475081  0.001469    0.875000  1.841790  0.000671    4.199349  
11            0.475081  0.001364    0.866667  1.824249  0.000616    3.936890  
8             0.475081  0.001259    0.857143  1.804202  0.000561    3.674431  
5             0.475081  0.001049    0.833333  1.754086  0.000451    3.149512  
12            0.475081  0.001574    0.833333  1.754086  0.000677    3.149512  
0             0.475081  0.003253    0.815789  1.717158  0.001358    2.849559  
1             0.475081  0.005351    0.809524  1.703969  0.002211    2.755823  
6             0.475081  0.001574    0.789474  1.661765  0.000627    2.493364  
9             0.475081  0.001784    0.772727  1.626516  0.000687    2.309642  
3             0.475081  0.001049    0.769231  1.619156  0.000401    2.274648

In [309]:
# lift  = support(Toast And Coffee)/Support(Toast | coffee)
## Just checking Lift for one of item
print(f"Lift : {0.001469/(0.001679*0.475081)}")


Lift : 1.8416344811162066


## Conclusion
 * it has been observed that:
 * 'Salad', 'Extra Salami or Feta' brought 0.16% of all the transcaction
 * Coffee has been brought 47.5% of all the transcaction
 * 'Salad', 'Extra Salami or Feta' and Coffee has been brought together with confidence 87.5%
 * 'Salad', 'Extra Salami or Feta' and Coffee is strongly associate with highest lift of 1.84

### Reference 
* "https://github.com/viktree/curly-octo-chainsaw/blob/master/BreadBasket_DMS.csv"
* https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&ved=2ahUKEwj-4qW15J_qAhUwzTgGHQ5MCuUQFjAHegQICRAB&url=http%3A%2F%2Fwww.ijarcs.info%2Findex.php%2FIjarcs%2Farticle%2Fdownload%2F4564%2F4083&usg=AOvVaw0tJaQUepruvpCogDKbi7T3